Stack removal: platform changes (#335)
buildpacks/spec: platform.md

  - [Additional Guidance](#additional-guidance)
    - [Environment](#environment)
      - [Buildpack Environment](#buildpack-environment)

>         - [Base Image-Provided Variables](#base-image-provided-variables)

Author: Updated in 9957239

#### CNB Terminology
A **buildpack** refers to software compliant with the [Buildpack Interface Specification](buildpack.md).

> A **stack** is a contract, implemented by a **build image** and **run image**, that guarantees properties of the **build environment** and **app image**.

Author: Maybe we should leave this in terminology but call it out as being deprecated

> A **stack** is a contract, implemented by a **build image** and **run image**, that guarantees properties of the **build environment** and **app image**.

Author: Added back `stack` in a2d78bb

#### CNB Terminology
A **buildpack** refers to software compliant with the [Buildpack Interface Specification](buildpack.md).

> A **base image** is an OCI image containing the base, or initial set of layers, for other images.

Reviewer: Shouldn't the two lines below reference this definition? I.e., shouldn't they both say "… is an OCI image that ~provides~ *serves as* the base *image* of the …"?

Author: Thank you! I updated in 4f4aa24

- The image config's `Label` field has the label `io.buildpacks.base.homepage` set to the homepage of the image.
- The image config's `Label` field has the label `io.buildpacks.base.released` set to the release date of the image.
- The image config's `Label` field has the label `io.buildpacks.base.description` set to the description of the image.
- The image config's `Label` field has the label `io.buildpacks.base.metadata` set to additional metadata related to the image.

### Run Image

A typical run image might specify:
* The OS distro or distroless OS in the launch environment.
* OS packages installed in the launch environment.
* Trusted CA certificates in the launch environment.
* The default user in the run environment.

The platform MUST ensure that:

- The image config's `Env` field has the environment variable `PATH` set to a valid set of paths or explicitly set to empty (`PATH=`).

The platform SHOULD ensure that:

- The image config's `User` field is set to a user with a **DIFFERENT** user [†](README.md#operating-system-conventions)UID/[‡](README.md#operating-system-conventions)SID as the build image.

> - The image config's `Label` field has the label `io.buildpacks.id` set to the target ID (e.g., "minimal") of the image.

Author: Should this be `io.buildpacks.base.id` (for consistency with the other labels)?

Author: On second thought... no? Because `base` is something that could apply to a build OR a run image, and the "target id" is only for the run image.

- The image config's `Label` field has the label `io.buildpacks.base.maintainer` set to the name of the image maintainer.
- The image config's `Label` field has the label `io.buildpacks.base.homepage` set to the homepage of the image.
- The image config's `Label` field has the label `io.buildpacks.base.released` set to the release date of the image.
- The image config's `Label` field has the label `io.buildpacks.base.description` set to the description of the image.
- The image config's `Label` field has the label `io.buildpacks.base.metadata` set to additional metadata related to the image.

### Target Data

For both build images and run images, the platform MUST ensure that:

- The image config's `os` and `architecture` fields are set to valid identifiers as defined in the [OCI Image Specification](https://github.com/opencontainers/image-spec/blob/main/config.md).

> - The build image config and the run image config both specify the same `os`, `architecture`, `variant` (if specified), `io.buildpacks.distribution.name` (if specified), and `io.buildpacks.distribution.version` (if specified).

Author: We used to have `io.buildpacks.stack.distro.name` - do we like `distro` or `distribution`?

Reviewer 2: I think `distribution` is fine - but no strong preference.

The platform SHOULD ensure that:

- The image config's `variant` field is set to a valid identifier as defined in the [OCI Image Specification](https://github.com/opencontainers/image-spec/blob/main/config.md).
- The image config's `Label` field has the label `io.buildpacks.distribution.name` set to the OS distribution and the label `io.buildpacks.distribution.version` set to the OS distribution version.
  - For Linux-based images, each label should contain the values specified in `/etc/os-release` (`$ID` and `$VERSION_ID`), as the `os.version` field in an image config may contain combined distribution and version information.
  - For Windows-based images, `io.buildpacks.distribution.name` should be empty; `io.buildpacks.distribution.version` should contain the value of `os.version` in the image config (e.g., `10.0.14393.1066`).

### Compatibility Guarantees

> - Base image authors SHOULD ensure that build image versions maintain [ABI-compatibility](https://en.wikipedia.org/wiki/Application_binary_interface) with previous versions, although violating this requirement will not change the behavior of previously built images containing app and launch layers.

Reviewer 3: Is there a reason this is `SHOULD` vs. `MUST`. I'm just trying to think through the implications. If the ABI compatibility changes, but it's not known to the Buildpack Author, how will they know if they need to invalidate the buildpack cache? In the past they could use the `CNB_STACK_ID` for this purpose and store it as layer metadata.

Author: @samj1912 do you have an opinion here?

Author: I feel mixed about this - would the stack ID have always changed previously if there were changes to ABI compat? We seem to have relied on mixins for this purpose.

Reviewer 3: @natalieparellano I'm not sure I understand what you're saying. Are you saying we only demanded ABI compatibility for mixins but not in the build? I'm looking at the [spec compat guarantees](https://github.com/buildpacks/spec/blob/main/platform.md#compatibility-guarantees) where this phrasing comes from:

> Stack image authors SHOULD ensure that build image versions maintain [ABI-compatibility](https://en.wikipedia.org/wiki/Application_binary_interface) with previous versions, although violating this requirement will not change the behavior of previously built images containing app and launch layers.

What you have written seems to match that, but I do wonder about the use case on where we want to break ABI compatibility here.

Reviewer 3: > I feel mixed about this - would the stack ID have always changed previously if there were changes to ABI compat? We seem to have relied on mixins for this purpose.

That's fair. I think at Heroku, we've only cared about forward compatible changes in the stack for both the build + run image.

I do think it's worth discussing why build/run are different in the wording.

Reviewer 4: My impression (could easily be wrong):

- it was previously a “should” because mixins were a workaround
- but nobody really used mixins, so in practice it was more like a “really ought to because you’ll have other problems if you don’t”
- dockerfile extensions are still a potential workaround/backdoor that could create issues by creating an ABI incompatibility that isn’t reflected in the target ID.
- it probably doesn’t cost anyone anything to change this to a MUST, and it’s probably good advice / good practice to follow.

@natalieparellano et. al. wdyt about changing this to a MUST ?

Author: > why build/run are different in the wording

I think that's because of rebase. I guess in theory, if you change ABI compat for the build image, you can always also upgrade your buildpacks. 

I'm happy to change SHOULD -> MUST here if that's desired.

Author: > who is "you" in that case

You're right. In my head, it was the "platform operator" who is presumably overseeing builder creation. It is fair to point out that the personas may be different depending on the organization.

Reviewer 3: I've made #368 to move the discussion to...

- The image config's `Env` field has the environment variable `CNB_USER_ID` set to the user [†](README.md#operating-system-conventions)UID/[‡](README.md#operating-system-conventions)SID of the user specified in the `User` field.
- The image config's `Env` field has the environment variable `CNB_GROUP_ID` set to the primary group [†](README.md#operating-system-conventions)GID/[‡](README.md#operating-system-conventions)SID of the user specified in the `User` field.
- The image config's `Env` field has the environment variable `PATH` set to a valid set of paths or explicitly set to empty (`PATH=`).

The platform SHOULD ensure that:

- The image config's `Label` field has the label `io.buildpacks.base.maintainer` set to the name of the image maintainer.
- The image config's `Label` field has the label `io.buildpacks.base.homepage` set to the homepage of the image.
- The image config's `Label` field has the label `io.buildpacks.base.released` set to the release date of the image.
- The image config's `Label` field has the label `io.buildpacks.base.description` set to the description of the image.

> - The image config's `Label` field has the label `io.buildpacks.base.metadata` set to additional metadata related to the image.

Author: These labels weren't mentioned in the RFC... I assume we still want something like them.

The platform MUST ensure that:

- The image config's `Env` field has the environment variable `PATH` set to a valid set of paths or explicitly set to empty (`PATH=`).

The platform SHOULD ensure that:

- The image config's `User` field is set to a user with a **DIFFERENT** user [†](README.md#operating-system-conventions)UID/[‡](README.md#operating-system-conventions)SID as the build image.
- The image config's `Label` field has the label `io.buildpacks.id` set to the target ID (e.g., "minimal") of the image.
- The image config's `Label` field has the label `io.buildpacks.base.maintainer` set to the name of the image maintainer.
- The image config's `Label` field has the label `io.buildpacks.base.homepage` set to the homepage of the image.
- The image config's `Label` field has the label `io.buildpacks.base.released` set to the release date of the image.
- The image config's `Label` field has the label `io.buildpacks.base.description` set to the description of the image.
- The image config's `Label` field has the label `io.buildpacks.base.metadata` set to additional metadata related to the image.

> ### Platform Data

Author: There could probably be a better header name for this

- The image config's `Env` field has the environment variable `CNB_GROUP_ID` set to the primary group [†](README.md#operating-system-conventions)GID/[‡](README.md#operating-system-conventions)SID of the user specified in the `User` field.
- The image config's `Env` field has the environment variable `PATH` set to a valid set of paths or explicitly set to empty (`PATH=`).

The platform SHOULD ensure that:

- The image config's `Label` field has the label `io.buildpacks.base.maintainer` set to the name of the image maintainer.
- The image config's `Label` field has the label `io.buildpacks.base.homepage` set to the homepage of the image.
- The image config's `Label` field has the label `io.buildpacks.base.released` set to the release date of the image.
- The image config's `Label` field has the label `io.buildpacks.base.description` set to the description of the image.
- The image config's `Label` field has the label `io.buildpacks.base.metadata` set to additional metadata related to the image.

### Run Image

> A typical run image might specify:

Reviewer 4: probably a nit:  in this sentence what's the difference between "specify" and "include"?     I think of "specify" as meaning "i'm describing something with a declarative config, so i'm _specifying_ the desired end state",  but I don't think that's how run images work --  run images are the image artifact, right?  so it just kinda "has stuff"?  I think the formal way of saying "has stuff" is "includes" but anyway I could be wrong in several different ways here.

Author: specify -> determine in a2d78bb

- The image config's `Env` field has the environment variable `PATH` set to a valid set of paths or explicitly set to empty (`PATH=`).

The platform SHOULD ensure that:

- The image config's `User` field is set to a user with a **DIFFERENT** user [†](README.md#operating-system-conventions)UID/[‡](README.md#operating-system-conventions)SID as the build image.
- The image config's `Label` field has the label `io.buildpacks.id` set to the target ID (e.g., "minimal") of the image.
- The image config's `Label` field has the label `io.buildpacks.base.maintainer` set to the name of the image maintainer.
- The image config's `Label` field has the label `io.buildpacks.base.homepage` set to the homepage of the image.
- The image config's `Label` field has the label `io.buildpacks.base.released` set to the release date of the image.
- The image config's `Label` field has the label `io.buildpacks.base.description` set to the description of the image.
- The image config's `Label` field has the label `io.buildpacks.base.metadata` set to additional metadata related to the image.

### Target Data

For both build images and run images, the platform MUST ensure that:

> - The image config's `os` and `architecture` fields are set to valid identifiers as defined in the [OCI Image Specification](https://github.com/opencontainers/image-spec/blob/main/config.md).

Reviewer 4: The OCI image spec uses "SHOULD" (as opposed to must) and defers to the go compiler's GOARCH list for valid architecture values.  However, we have a use-case for `*` architectures which doesn't exist in the go-compiler (example: a buildpack that is just shell scripts or batch scripts, which are inherently multi-arch).  Can we ammend this line to indicate that we're comfortable with a superset of the OCI spec's valid Architectures that also includes `*` arch?

Author: Does `*` arch mean missing arch, or literally `*`?

Reviewer 4: I think it really means "any"

Author: Ah, now I understand (maybe). If we're talking about matching a buildpack that has stack `*` in buildpack.toml, then we might want to update the language around backwards-compat in the buildpack spec. I made a code suggestion here: https://github.com/buildpacks/spec/pull/336/files#r1142629644 - WDYT?

- The image config's `Label` field has the label `io.buildpacks.base.homepage` set to the homepage of the image.
- The image config's `Label` field has the label `io.buildpacks.base.released` set to the release date of the image.
- The image config's `Label` field has the label `io.buildpacks.base.description` set to the description of the image.
- The image config's `Label` field has the label `io.buildpacks.base.metadata` set to additional metadata related to the image.

### Run Image

A typical run image might determine:
* The OS distro or distroless OS in the launch environment.
* OS packages installed in the launch environment.
* Trusted CA certificates in the launch environment.
* The default user in the run environment.

The platform MUST ensure that:

- The image config's `Env` field has the environment variable `PATH` set to a valid set of paths or explicitly set to empty (`PATH=`).

The platform SHOULD ensure that:

- The image config's `User` field is set to a user with a **DIFFERENT** user [†](README.md#operating-system-conventions)UID/[‡](README.md#operating-system-conventions)SID as the build image.

> - The image config's `Label` field has the label `io.buildpacks.id` set to the target ID (e.g., "minimal") of the image.

Author: ```suggestion
- The image config's `Label` field has the label `io.buildpacks.id` set to the target ID (e.g., "minimal") of the run image.
```

> - The image config's `Label` field has the label `io.buildpacks.id` set to the target ID (e.g., "minimal") of the run image.

Reviewer 3: what is "minimal" here? (I'm not understanding the example).

Reviewer 4: should it be tiny/base/full [as in paketo](https://hub.docker.com/r/paketobuildpacks/builder)?

Reviewer 3: @natalieparellano do we use `io.buildpacks.run` anywhere today? Why `run` over `base`?

Author: To my knowledge, `io.buildpacks.run` isn't used anywhere today. I like it better than `base` because base is ambiguous - which base? The build-time base, or runtime base, or both? IMO it leads to an ambiguous coupling between the build- and runtime base images which we are trying to avoid with the removal of the concept of "stack". Though, to be fair, they are always going to be tied together to some degree. I won't die on this hill, just putting it out there...

Reviewer 3: I don't have a strong preference besides we aren't doing anything with `io.buildpacks.run` today. Are there other things we should put there or would you like to just start with this? What I do care about is it remaining on `io.buildpacks.id`, so as long as we change that, I'm :+1: .

Author: Or how about `io.buildpacks.target.id` (for parity with `CNB_TARGET_ID`)? Or `io.buildpacks.target`?

Author: @hone I don't feel strongly about this. It's hard to provide guidance about a feature that's not being used yet. I think in the future we might be able to give some more useful examples.

Reviewer 3: @natalieparellano I think the counterpoint, we're adding this field for a reason and with intention. From what I seem to be gathering so far is it being used in place of `CNB_STACK_ID`. Especially, since we're providing this to Buildpack Authors in the Buildpack API, Buildpacks can use it to change logic based on this value. From WG yesterday, what happens if there's different base/builder images that use the same target id?

Reviewer 3: Looking at other parts of the spec, when we use "identifier", it usually is "machine readable" and has restrictions on what we allow in it:

* MUST NOT be identical to any other ID when using a case-insensitive comparison.
* MUST only contain numbers, letters, and the characters ., /, and -.
* SHOULD use reverse domain name notation to avoid name collisions.

While we use "name" to have something human readable without any restrictions.

Does it make sense to apply similar constraints here?

Author: Sounds good - please see bc1e753

Reviewer: Cool. I've further updated some of my thoughts on the target ID over in https://github.com/buildpacks/community/discussions/230; maybe we can discuss in the WG meeting on Thursday?

Reviewer 3: This is great. Thanks for working with me on this. We can do the rest of the guidance I'm looking for on "what" to use target id for on the documentation side.

> - The image config's `Label` field has the label `io.buildpacks.id` set to the target ID (e.g., "minimal") of the run image.

Author: ```suggestion
- The image config's `Label` field has the label `io.buildpacks.base.id` set to the target ID (e.g., "minimal") of the run image.
```

- The image config's `Label` field has the label `io.buildpacks.base.maintainer` set to the name of the image maintainer.
- The image config's `Label` field has the label `io.buildpacks.base.homepage` set to the homepage of the image.
- The image config's `Label` field has the label `io.buildpacks.base.released` set to the release date of the image.
- The image config's `Label` field has the label `io.buildpacks.base.description` set to the description of the image.
- The image config's `Label` field has the label `io.buildpacks.base.metadata` set to additional metadata related to the image.

### Target Data

For both build images and run images, the platform MUST ensure that:

- The image config's `os` and `architecture` fields are set to valid identifiers as defined in the [OCI Image Specification](https://github.com/opencontainers/image-spec/blob/main/config.md).

> - The build image config and the run image config both specify the same `os`, `architecture`, `variant` (if specified), `io.buildpacks.distribution.name` (if specified), and `io.buildpacks.distribution.version` (if specified).

Author: ```suggestion
- The build image config and the run image config both specify the same `os`, `architecture`, `variant` (if specified), `io.buildpacks.base.distro.name` (if specified), and `io.buildpacks.base.distro.version` (if specified).
```

Reviewer: Is there a particular reason for changing `distribution` to `distro`? Would it not be better to remain consistent with e.g. `target.distribution`?

Author: Ah, do you mean the `[[targets]]` table in buildpack.toml? I guess there could be an argument for shortening that to `distro` as well. We already provide `CNB_TARGET_DISTRO_NAME` in the buildpack environment to indicate the target.

Here, we changed distribution -> distro because namespacing under `base` made the label very long.

Reviewer 3: I'm :+1: for consistency if we're going to do that.

The platform SHOULD ensure that:

- The image config's `variant` field is set to a valid identifier as defined in the [OCI Image Specification](https://github.com/opencontainers/image-spec/blob/main/config.md).

> - The image config's `Label` field has the label `io.buildpacks.distribution.name` set to the OS distribution and the label `io.buildpacks.distribution.version` set to the OS distribution version.

Author: ```suggestion
- The image config's `Label` field has the label `io.buildpacks.base.distro.name` set to the OS distribution and the label `io.buildpacks.base.distro.version` set to the OS distribution version.
```

  - For Linux-based images, each label should contain the values specified in `/etc/os-release` (`$ID` and `$VERSION_ID`), as the `os.version` field in an image config may contain combined distribution and version information.

>   - For Windows-based images, `io.buildpacks.distribution.name` should be empty; `io.buildpacks.distribution.version` should contain the value of `os.version` in the image config (e.g., `10.0.14393.1066`).

Author: ```suggestion
  - For Windows-based images, `io.buildpacks.base.distro.name` should be empty; `io.buildpacks.base.distro.version` should contain the value of `os.version` in the image config (e.g., `10.0.14393.1066`).
```

> The meaning of any particular instance of the word **build** must be assessed in context

#### Rebase

> When an updated run image is available, an updated app image SHOULD be generated from the existing app image config by replacing the run image layers in the existing app image with the layers from the new run image.

Reviewer 3: Do we care that the distribution name/versions match or is that covered by the "MUST" have ABI compatibility here? Another thought, should `io.buildpacks.id` match?

Usage:
```
/cnb/lifecycle/detector \
  [-analyzed <analyzed>] \

>   [-app <app>] \

Author: We'll need to add `-analyzed` to the builder as well to provide target information during the build phase

Author: Added in 1fe9fd4

| `<log-level>`   | `CNB_LOG_LEVEL`      | `info`                   | Log Level                                                                   |
| `<uid>`         | `CNB_USER_ID`        |                          | UID of the build image `User`                                               |
| `<skip-layers>` | `CNB_SKIP_LAYERS`    | `false`                  | Do not perform [layer restoration](#layer-restoration)                      |

> | `<kaniko-dir>`  |                      |                          | Kaniko directory (must be `/kaniko`)                                        |

Author: Unrelated change... I did and undid something

Usage:
```
/cnb/lifecycle/builder \

>   [-analyzed <analyzed>] \

Author: Needed to provide target data to buildpacks

```

##### Inputs
| Input            | Env                    | Default Value            | Description                                                                                    |
|------------------|------------------------|--------------------------|------------------------------------------------------------------------------------------------|

> | `<analyzed>`     | `CNB_ANALYZED_PATH`    | `<layers>/analyzed.toml` | Path to analysis metadata (see [`analyzed.toml`](#analyzedtoml-toml)                           |

Author: Added analyzed

| `<group>`                       | `CNB_GROUP_PATH`            | `<layers>/group.toml`            | Path to group file (see [`group.toml`](#grouptoml-toml))                                   |
| `<image>`                       |                             |                                  | Tag reference to which the app image will be written                                       |
| `<launch-cache>`                | `CNB_LAUNCH_CACHE_DIR`      |                                  | Path to a cache directory containing launch layers                                         |
| `<launcher-sbom>`               |                             | `/cnb/lifecycle`                 | Path to directory containing SBOM files describing the `launcher` executable               |
| `<launcher>`                    |                             | `/cnb/lifecycle/launcher`        | Path to the `launcher` executable                                                          |

> | `<layers>/config/metadata.toml` |                             |                                  | Build metadata (see [`metadata.toml`](#metadatatoml-toml)                                  |

Author: Alphabetized

```

##### Inputs
| Input         | Environment Variable | Default Value          | Description                                              |
|---------------|----------------------|------------------------|----------------------------------------------------------|
| `<daemon>`    | `CNB_USE_DAEMON`     | `false`                | Export image to docker daemon                            |

> | `<force>`     | `CNB_FORCE_REBASE`   | `false`                | Allow mismatched metadata between new and old run images |

Author: Reformatted table and added `<force>`

| `<gid>`       | `CNB_GROUP_ID`       |                        | Primary GID of the build image `User`                    |
| `<image>`     |                      |                        | App image to rebase                                      |
| `<log-level>` | `CNB_LOG_LEVEL`      | `info`                 | Log Level                                                |
| `<report>`    | `CNB_REPORT_PATH`    | `<layers>/report.toml` | Path to report (see [`report.toml`](#reporttoml-toml)    |
| `<run-image>` | `CNB_RUN_IMAGE`      | derived from `<image>` | Run image reference                                      |
| `<uid>`       | `CNB_USER_ID`        |                        | UID of the build image `User`                            |

- At least one `<image>` must be provided
- Each `<image>` MUST be a valid tag reference
- **If** `<daemon>` is `false` and more than one `<image>` is provided, the images MUST refer to the same registry

> - **If** `<run-image>` is not provided by the platform, the value will be [resolved](#run-image-resolution) from the contents of the `stack` key in the `io.buildpacks.lifecycle.metdata` label on `<image>`

Author: We should probably change the key name - maybe something like `base`?

Similarly for stack.toml - maybe something like run.toml?

Suggestions are welcome...

Author: Or we could use this opportunity to refactor the `io.buildpacks.lifecycle.metadata` label, which I know @ekcasey has long wanted ;)

Author: Summarizing offline discussions: let's keep it as `stack` for now. The `io.buildpacks.lifecycle.metdata` is "user facing" but not as user facing as the `io.buildpacks.stack.*` labels.

We may make further changes when we do run image extensions.

Author: Maybe instead of:
```
  "runImage": {
    "topLayer": "<run-image-top-layer-diffID>",
    "reference": "<run-image-reference>"
  },
  "stack": {
    "runImage": {
      "image": "cnbs/sample-stack-run:bionic"
    }
  }
```
in the label, we could have
```
  "runImage": {
    "image": "cnbs/sample-stack-run:bionic",
    "mirrors": ["<run-image-mirror>"],
    "digest": "<run-image-reference>",
    "topLayer": "<run-image-top-layer-diffID>",
  },
```
and promote `run-image` and `build-image` to top-level keys in builder.toml.

Author: This is addressed in https://github.com/buildpacks/spec/pull/347

```

##### Inputs

| Input              | Environment Variable | Default Value          | Description                                              |
|--------------------|----------------------|------------------------|----------------------------------------------------------|
| `<daemon>`         | `CNB_USE_DAEMON`     | `false`                | Export image to docker daemon                            |

> | `<force>`          | `CNB_FORCE_REBASE`   | `false`                | Allow mismatched metadata between new and old run images |

Author: Added force

    - The value of `io.buildpacks.lifecycle.metadata` SHALL be modified as follows
      - `run-image.reference` SHALL uniquely identify `<run-image>`
      - `run-image.top-layer` SHALL be set to the uncompressed digest of the top layer in `<run-image>`

>     - The value of `io.buildpacks.base.*` labels SHALL be modified to that of the new `run-image`

Author: Should we also update `io.buildpacks.stack.*` labels if they are present? For backwards compat?

>     - The value of `io.buildpacks.base.*` labels SHALL be modified to that of the new `run-image`

Author: ```suggestion
    - The value of `io.buildpacks.base.*` labels (and any `io.buildpacks.stack.*` labels, if present) SHALL be modified to that of the new `run-image`
```

    - The value of `io.buildpacks.lifecycle.metadata` SHALL be modified as follows
      - `run-image.reference` SHALL uniquely identify `<run-image>`
      - `run-image.top-layer` SHALL be set to the uncompressed digest of the top layer in `<run-image>`
    - The value of `io.buildpacks.base.*` labels and `io.buildpacks.stack.*` labels (if present) SHALL be modified to that of the new `run-image`
- **If** `<force>` is `true`, the following [target data](#target-data) values in the output `<image>` config MUST be derived from the new `<run-image>`:
  - `os`
  - `architecture`
  - `variant` (if specified)
  - `io.buildpacks.distribution.name` (if specified)
  - `io.buildpacks.distribution.version` (if specified)

> - **Else** they MUST match the old run image if `<force>` is `false`

Reviewer 3: by "they" I assume you mean the target data above?

> - **Else** they MUST match the old run image if `<force>` is `false`

Author: ```suggestion
- **Else** the target data above MUST match the old run image if `<force>` is `false`
```

- **If** `<force>` is `true` and the provided `<run-image>` is not found in `runImage.image` or `runImage.mirrors`:
  - `run-image.image` SHALL be the provided `<run-image>`
  - `run-image.mirrors` SHALL be omitted

> - **Else** the provided `<run-image>` MUST be found in `runImage.image` or `runImage.mirrors` if `<force>` is `false`

Reviewer 3: Does it make sense to make this an `**Else If**`?

```suggestion
- **Else If** `<force> is `false`, the provided `<run-image>` MUST be found in `runImage.image` or `runImage.mirrors`
```

Author: Updated capitalization, but yes this reads better 🙏🏼

      - `run-image.reference` SHALL uniquely identify `<run-image>`
      - `run-image.top-layer` SHALL be set to the uncompressed digest of the top layer in `<run-image>`
    - The value of `io.buildpacks.stack.*` labels SHALL be modified to that of the new `run-image`
- **If** the provided `<run-image>` is not found in `runImage.image` or `runImage.mirrors`:
      - `run-image.image` SHALL be the provided `<run-image>`

>       - `run-image.mirrors` SHALL be omitted

Author: Changes from https://github.com/buildpacks/spec/pull/360

Author: We need to update the previous line to also include `io.buildpacks.base.*`

### Run Image Resolution

Given [run](#runtoml-toml) metadata, the `<run-image>` for a given `<image>` shall be resolved as follows:

> - By choosing an image from `[[images]]`

Author: ```suggestion
- By choosing data for labeling the destination image
```

  - **If** the tag reference for the desired run image is known (e.g., from a `-run-image` flag or `analyzed.toml`):
    - The first image in `[[images]]` where `image.image` or one of `image.mirrors` has a matching tag reference
  - **Else** the first image in `[[images]]`

> - By choosing the best mirror for an image

Author: ```suggestion
- By choosing the run image when the destination image is known
```

[run-image]
  reference = "<image reference>"

>   target-id = "<target identifer>"

Author: Question: should this be `target.id`?

>   target-id = "<target identifer>"
>   [target]

Author: ```suggestion
  [target]
  id = "<target identifer>"
```

  os = "<OS name>"
  arch = "<architecture>"
  variant = "<architecture variant>"
  [target.distribution]
  name = "<OS distribution name>"
  version = "<OS distribution version>"

[build-image]
  reference = "<image reference>"
```

Where:
- `image.reference` MUST be either a digest reference to an image in an OCI registry or the ID of an image in a docker daemon
- `image.metadata` MUST be the TOML representation of the layer [metadata label](#iobuildpackslifecyclemetadata-json)
- `run-image.reference` MUST be either a digest reference to an image in an OCI registry or the ID of an image in a docker daemon

> - `run-image.target-id` is optional and MUST be the value of the label `io.buildpacks.id`

Author: ```suggestion
- `run-image.target.id` is optional and MUST be the value of the label `io.buildpacks.id`
```

[run-image]
  reference = "<image reference>"
  [target]
  id = "<target identifer>"
  os = "<OS name>"
  arch = "<architecture>"
  variant = "<architecture variant>"
  [target.distribution]
  name = "<OS distribution name>"
  version = "<OS distribution version>"

[build-image]
  reference = "<image reference>"
```

Where:
- `image.reference` MUST be either a digest reference to an image in an OCI registry or the ID of an image in a docker daemon
- `image.metadata` MUST be the TOML representation of the layer [metadata label](#iobuildpackslifecyclemetadata-json)
- `run-image.reference` MUST be either a digest reference to an image in an OCI registry or the ID of an image in a docker daemon
- `run-image.target.id` is optional and MUST be the value of the label `io.buildpacks.id`
- `run-image.target` contains the [target data](#target-data) for the image
  - If target data is missing but the image contains the label `io.buildpacks.stack.id` with value `io.buildpacks.stacks.bionic`, the lifecycle SHALL assume the following values:
    - `run-image.target.os = "linux"`

>     - `run-image.target.arch = "x86_64"`

Reviewer 5: Do we want to use values here that match `go tool dist list`?

Reviewer 3: ```suggestion
    - `run-image.target.arch = "arm64"`
```
yes. I think that's derived from [$GOARCH](https://go.dev/doc/install/source#environment). Per the [RFC](https://github.com/buildpacks/rfcs/blob/main/text/0096-remove-stacks-mixins.md), we should align with [OCI Image Config Spec](https://github.com/opencontainers/image-spec/blob/main/config.md):

> architecture string, REQUIRED
>
> The CPU architecture which the binaries in this image are built to run on. Configurations SHOULD use, and implementations SHOULD understand, values listed in the Go Language document for [GOARCH](https://golang.org/doc/install/source#environment).

[run-image]
  reference = "<image reference>"
  [target]
  id = "<target identifer>"
  os = "<OS name>"
  arch = "<architecture>"
  variant = "<architecture variant>"
  [target.distribution]
  name = "<OS distribution name>"
  version = "<OS distribution version>"

[build-image]
  reference = "<image reference>"
```

Where:
- `image.reference` MUST be either a digest reference to an image in an OCI registry or the ID of an image in a docker daemon
- `image.metadata` MUST be the TOML representation of the layer [metadata label](#iobuildpackslifecyclemetadata-json)
- `run-image.reference` MUST be either a digest reference to an image in an OCI registry or the ID of an image in a docker daemon
- `run-image.target.id` is optional and MUST be the value of the label `io.buildpacks.id`
- `run-image.target` contains the [target data](#target-data) for the image

>   - If target data is missing but the image contains the label `io.buildpacks.stack.id` with value `io.buildpacks.stacks.bionic`, the lifecycle SHALL assume the following values:
>     - `run-image.target.os = "linux"`
>     - `run-image.target.arch = "arm64"`
>     - `run-image.target.distribution.name = "ubuntu"`
>     - `run-image.target.distribution.version = "18.04"`

Author: ```suggestion
  - If target distribution data is missing, it will be inferred from `/etc/os-release` for Linux images; furthermore, if the image contains the label `io.buildpacks.stack.id` with value `io.buildpacks.stacks.bionic`, the lifecycle SHALL assume the following values:
    - `run-image.target.os = "linux"`
    - `run-image.target.arch = "arm64"`
    - `run-image.target.distribution.name = "ubuntu"`
    - `run-image.target.distribution.version = "18.04"`
```

[run-image]
  reference = "<image reference>"
  [target]
  id = "<target identifer>"
  os = "<OS name>"
  arch = "<architecture>"
  variant = "<architecture variant>"
  [target.distribution]
  name = "<OS distribution name>"
  version = "<OS distribution version>"

[build-image]
  reference = "<image reference>"
```

Where:
- `image.reference` MUST be either a digest reference to an image in an OCI registry or the ID of an image in a docker daemon

> - `image.metadata` MUST be the TOML representation of the layer [metadata label](#iobuildpackslifecyclemetadata-json)

Author: Fixes typo in existing spec

  os = "<OS name>"
  arch = "<architecture>"
  variant = "<architecture variant>"
  [target.distribution]
  name = "<OS distribution name>"
  version = "<OS distribution version>"

[build-image]
  reference = "<image reference>"
```

Where:
- `image.reference` MUST be either a digest reference to an image in an OCI registry or the ID of an image in a docker daemon
- `image.metadata` MUST be the TOML representation of the layer [metadata label](#iobuildpackslifecyclemetadata-json)
- `run-image.reference` MUST be either a digest reference to an image in an OCI registry or the ID of an image in a docker daemon
- `run-image.target.id` is optional and MUST be the value of the label `io.buildpacks.id`
- `run-image.target` contains the [target data](#target-data) for the image
  - If target distribution data is missing, it will be inferred from `/etc/os-release` for Linux images; furthermore, if the image contains the label `io.buildpacks.stack.id` with value `io.buildpacks.stacks.bionic`, the lifecycle SHALL assume the following values:
    - `run-image.target.os = "linux"`
    - `run-image.target.arch = "arm64"`
    - `run-image.target.distribution.name = "ubuntu"`

>     - `run-image.target.distribution.version = "18.04"`

Reviewer 3: I assume this means if for some reason if target distribution data is missing + `/etc/os-release` doesn't exist these will be blank? Will we try to do something based on the host system?

Reviewer 4: IIRC checking the os-release file was the best we could think of to do with respect to the host system.  TBH it "shouldn't" usually be blank, and while i'm sure there's cases, even e.g. the smallest "distroless" image (distroless/static) still has that file.

Reviewer 4: but yes,  if that information can't be found anywhere, then those values would be blank.

Reviewer: FWIW, from https://www.linux.org/docs/man5/os-release.html:

> The file `/etc/os-release` takes precedence over `/usr/lib/os-release`. Applications should check for the former,
> and exclusively use its data if it exists, and only fall back to `/usr/lib/os-release` if it is missing.
> Applications should not read data from both files at the same time.  `/usr/lib/os-release` is the recommended
> place to store OS release information as part of vendor trees.  `/etc/os-release` should be a relative symlink
> to `/usr/lib/os-release`, to provide compatibility with applications only looking at `/etc`. A relative symlink
> instead of an absolute symlink is necessary to avoid breaking the link in a `chroot` or `initrd` environment such
> as dracut.

}
```
This label MUST contain the JSON representation of [`project-metadata.toml`](#project-metadatatoml-toml)

## Deprecations
This section describes all the features that are deprecated.

### `io.buildpacks.stack.*` Labels

_Deprecated in Platform API 0.12._

For compatibility with older platforms and older buildpacks, base image authors SHOULD ensure for build images and run images:

- The image config's `Env` field has the environment variable `CNB_STACK_ID` set to the stack ID.
- The image config's `Label` field has the label `io.buildpacks.stack.id` set to the stack ID.
- The image config's `Label` field has the label `io.buildpacks.stack.mixins` set to a JSON array containing mixin names for each mixin applied to the image.

To upgrade, the platform SHOULD upgrade all buildpacks to use Buildpack API `0.10` or greater.

> ### `io.buildpacks.lifecycle.metadata` (JSON) `stack` Key

Reviewer 4: i guess to my question from above,  does it make sense to move most of the deleted stuff about stacks down into this section and basically say like, 

> the lifecycle SHOULD prefer the spec above but MUST fall back on the deprecated stack logic until [ the heat death of the universe | buildpacks 1.0 | end of calendar year 2024 | we say so] ?

Author: Yeah... although, I would prefer to only add the stuff that falls within the responsibility of the lifecycle. So stuff like:
> The image config's `Label` field has the label `io.buildpacks.stack.maintainer` set to the name of the stack maintainer.

I would prefer to leave off because it really makes no difference to the lifecycle if that label is set or not. Mixins are kind of in a similar boat, even though somewhere I think we say that the lifecycle validates mixins, it does not. That responsibility is left up to the platform. If a platform wants to support earlier platform API versions, it can consult the `platform/0.X` version of the spec.

Other two points might be worth moving or repeating here:
> **If** `<run-image>` is not provided by the platform, the value will be [resolved](#run-image-resolution) from the contents of the `runImage` key in the `io.buildpacks.lifecycle.metdata` label on `<image>`, or `stack.runImage` if not found (for compatibility with older platforms).

> | `CNB_STACK_ID` | Chosen stack ID

This variable will be passed-through by the lifecycle to buildpacks regardless of the buildpack API in use (in the current implementation).

I'm a bit mixed on moving everything into this section vs clearly marking things as deprecated where they are mentioned. We have preferred the former but it makes understanding deprecations a bit confusing.
